### PR TITLE
Implement OpAMP Client Basics

### PR DESCRIPTION
**Pull request by tigrannajaryan**
Resolves: https://github.com/open-telemetry/opamp-go/issues/17
Resolves: https://github.com/open-telemetry/opamp-go/issues/10
Resolves: https://github.com/open-telemetry/opamp-go/issues/8
Resolves: https://github.com/open-telemetry/opamp-go/issues/11

Implemented the following Client features:
- Connection establishment, retries, backoff.
- Status reporting.
- Remote configuration receiving.
- Connection settings receiving.

### CODE REVIEW

File: client/clientimpl.go [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+		logger:        logger,
+		stoppedSignal: make(chan struct{}, 1),
+		sender:        internal.NewSender(logger),
+	}
+	return w
+}
+
+func (w *Impl) Start(settings StartSettings) error {
+	if w.isStarted {
+		return errors.New("already started")
```
Comment (pmm-sumo) [inline client/clientimpl.go:74]: This is really a nitty nit, but perhaps all errors could be defined at the top of the file and referenced later?

File: client/clientimpl.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+
+	w.startConnectAndRun()
+
+	w.isStarted = true
+
+	return nil
+}
+
+func (w *Impl) Stop(ctx context.Context) error {
+	if !w.isStarted {
```
Comment (pmm-sumo) [inline client/clientimpl.go:119]: Should we assume that Start/Stop are thread-safe? If so, perhaps `isStarted` should be guarded or atomic variable could be used here? (then, maybe it's an overkill)

File: client/clientimpl.go [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+			statusReport.AgentDescription = &protobufs.AgentDescription{
+				AgentType:    w.settings.AgentType,
+				AgentVersion: w.settings.AgentVersion,
+			}
+
+			statusReport.ServerProvidedAllAddonsHash = w.settings.LastServerProvidedAllAddonsHash
+		},
+	)
+
+	w.startConnectAndRun()
```
Comment (pmm-sumo) [inline client/clientimpl.go:111]: Not sure if it's feasible, but if `Start()` would be called twice at the same time (e.g. from two threads) it could execute this code more than once. Don't think this would be common but perhaps we should guard with `isStarting` or so?

File: client/clientimpl.go [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+		retryAfter := strings.TrimSpace(resp.Header.Get(retryAfterHTTPHeader))
+		if retryAfter != "" {
+			retryIntervalSec, err := strconv.Atoi(retryAfter)
+			if err == nil {
+				retryInterval := time.Duration(retryIntervalSec) * time.Second
+				return retryInterval
+			}
+		}
+	}
+	return 0
```
Comment (pmm-sumo) [inline client/clientimpl.go:207]: I think that should be fine since `ExponentialBackoff` is used later on, but just wanted to check if that was the intention (I imagine that we could also consider returning the recommended value - 30s - as a default one rather than 0)

File: client/internal/receiver.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		r.processErrorResponse(err)
+	}
+}
+
+func (r *Receiver) rcvDataForAgent(data *protobufs.DataForAgent) {
+	if r.callbacks != nil {
+		reportStatus := false
+
+		reportStatus = r.rcvRemoteConfig(data.RemoteConfig) || reportStatus
+		reportStatus = r.rcvConnectionSettings(data.ConnectionSettings) || reportStatus
```
Comment (pmm-sumo) [inline client/internal/receiver.go:81]: `rcvConnectionSetting` will never return `true` though, will it?

File: client/internal/receiver.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	}
+}
+
+func (r *Receiver) rcvOtherConnectionSettings(
+	settings *protobufs.ConnectionSettings,
+	name string,
+	callback func(ctx context.Context, name string, settings *protobufs.ConnectionSettings) error,
+) {
+	if settings != nil {
+		callback(r.runContext, name, settings)
```
Comment (MovieStoreGuy) [inline client/internal/receiver.go:157]: I feel strongly against this, I think this would lead done some anti patterns and should be controlled in the `ReceiverLoop` by using the passed context.

**Review by MovieStoreGuy — COMMENTED** [review]
I had a done an initial review and there is a few things that I don't really agree but I am happy to hear your thoughts.

File: client/internal/callbacks.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
+package internal
+
+import (
+	"context"
+
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+type CallbacksStruct struct {
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply client/internal/callbacks.go:10]: I'm not opposed to this implementation. Usage of this struct is more intuitive to me than accessing functions in a map (e.g. `w.settings.Callbacks.OnConnectFailed(err)`).

File: client/clientimpl.go (hunk 3 above)
Comment (tigrannajaryan) [reply client/clientimpl.go:115]: Start() should not be called concurrently. I will add docs to Client interface to clarify this.

File: client/clientimpl.go (hunk 2 above)
Comment (tigrannajaryan) [reply client/clientimpl.go:123]: I will add docs to Client interface to clarify.

File: client/clientimpl.go [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+	if conn != nil {
+		conn.Close()
+	}
+	// If we are not connected by this point we may still get connected because
+	// we are racing with tryConnectOnce() func. However, after tryConnectOnce we will
+	// check the isStopping flag in receiverLoop() and will exit receiverLoop, so we will
+	// not deadlock.
+
+	// Wait until stopping is finished.
+	select {
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go:148]: Are you suggesting to make Stop() non-cancellable, so that it blocks indefinitely long until stopping is done?

File: client/clientimpl.go [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+func (w *Impl) ensureConnected(ctx context.Context) error {
+	infiniteBackoff := backoff.NewExponentialBackOff()
+
+	// Make ticker run forever.
+	infiniteBackoff.MaxElapsedTime = 0
+
+	interval := time.Duration(0)
+
+	for {
+		timer := time.NewTimer(interval)
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go:231]: That's a good point. I am not sure whether it is better to make the clock overridable or the backoff algo overridable. To keep the size of this PR manageable I suggest that we do it in a future PR which also adds more tests that involve failures and backoff.

File: client/clientimpl.go [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+					}
+
+					continue
+				}
+				// Connected successfully.
+				return nil
+			}
+
+		case <-ctx.Done():
+			w.logger.Infof("Client is stopped, will not try anymore.")
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go]: I agree. Changed to Debugf.

File: client/clientimpl.go [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+	defer w.isStoppingMutex.Unlock()
+
+	if w.isStoppingFlag {
+		// Stop() was called. Don't connect.
+		return
+	}
+	w.runCancel = runCancel
+	w.runContext = runCtx
+
+	go w.runUntilStopped()
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go]: It can't be already running since calling Start() twice is currently prohibited and is guarded against.

File: client/clientimpl.go [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+	for {
+		if w.isStopping() {
+			return
+		}
+
+		w.runOneCycle()
+	}
+}
+
+type nopLogger struct{}
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go]: > This should be moved to a different file to make discoverability easier.

Moved.

[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go]: >I suggest making part of types consider this would not be the only type using a no op logger.

I am not sure I understand, please clarify.

File: client/internal/callbacks.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/callbacks.go:10]: > This struct feels a little overkill and could potentially, to me this looks like a map of methods that can have values set?

> If that is the case, we can simplify with an enum and a map :)

I disagree. That would make impossible to make functions correctly strongly typed. With this approach every callback function has the appropriate signature with the right parameter  and return types.

File: client/internal/mockserver.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+	OnRequest func(w http.ResponseWriter, r *http.Request)
+	OnConnect func(r *http.Request, conn *websocket.Conn)
+	OnMessage func(msg *protobufs.AgentToServer) *protobufs.ServerToAgent
+	HTTPSrv   *http.Server
+}
+
+var upgrader = websocket.Upgrader{}
+
+func StartMockServer(t *testing.T) *MockServer {
+	srv := &MockServer{Endpoint: testhelpers.GetAvailableLocalAddress()}
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/mockserver.go:27]: I wasn't aware `httptest.NewServer` exists :-) Let me check.

File: client/internal/receiver.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+// Receiver implements the client's receiving portion of OpAMP protocol.
+type Receiver struct {
+	conn       *websocket.Conn
+	logger     types.Logger
+	sender     *Sender
+	runContext context.Context
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/receiver.go]: Removed.

File: client/internal/receiver.go [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+func (r *Receiver) receiveMessage(msg *protobufs.ServerToAgent) error {
+	_, bytes, err := r.conn.ReadMessage()
+	if err != nil {
+		return err
+	}
+	err = proto.Unmarshal(bytes, msg)
+	if err != nil {
+		r.logger.Errorf("Cannot decode received message: %v", err)
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/receiver.go]: Done (and wrapped).

File: client/internal/receiver.go (hunk 5 above)
Comment (tigrannajaryan) [reply client/internal/receiver.go]: Correct. Removed.

File: client/internal/receiver.go (hunk 6 above)
Comment (tigrannajaryan) [reply client/internal/receiver.go]: >I feel strongly against this

Against what specifically? Please clarify.

File: client/clientimpl.go [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gorilla/websocket"
+
+	"github.com/open-telemetry/opamp-go/client/internal"
+	"github.com/open-telemetry/opamp-go/client/types"
+	"github.com/open-telemetry/opamp-go/internal/protobufs"
+)
+
+// Impl is a Client implementation.
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go]: >Can we clarify whether this client is meant to be reusable and/or shareable?

It is not meant to be reusable. I added a clarifying comment to the interface. I am not sure what does "shareable" mean.

File: client/clientimpl.go [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+
+	// HTTP request headers to use when connecting to OpAMP Server.
+	requestHeader http.Header
+
+	// Websocket dialer and connection.
+	dialer    websocket.Dialer
+	conn      *websocket.Conn
+	connMutex sync.RWMutex
+
+	// True is Start() is successful.
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go]: Done.

File: client/clientimpl.go (hunk 1 above)
Comment (tigrannajaryan) [reply client/clientimpl.go]: I defined unexported errors at the top.

File: client/clientimpl.go (hunk 3 above)
Comment (tigrannajaryan) [reply client/clientimpl.go:115]: Clarified.

File: client/clientimpl.go (hunk 2 above)
Comment (tigrannajaryan) [reply client/clientimpl.go:123]: Clarified.

File: client/clientimpl.go (hunk 4 above)
Comment (tigrannajaryan) [reply client/clientimpl.go]: Refactored a bit to make more explicit that the interval is undefined. The caller takes care of the interval in that case.

File: client/internal/mockserver.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/mockserver.go]: Changed to `httptest.NewServer`, but saved virtually nothing compared to what it was before. Will keep it this way,  doesn't make a difference.

File: client/internal/mockserver.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+	m := http.NewServeMux()
+	m.HandleFunc(
+		"/", func(w http.ResponseWriter, r *http.Request) {
+			if srv.OnRequest != nil {
+				srv.OnRequest(w, r)
+				return
+			}
+
+			conn, err := upgrader.Upgrade(w, r, nil)
```
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/mockserver.go:39]: Why? What does that help with?

**Comment by tigrannajaryan** [comment]
@djaglowski @MovieStoreGuy @pmm-sumo thanks a lot for reviews. Please take another look.

**Review by pmm-sumo — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]

File: client/clientimpl.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply client/clientimpl.go]: My understanding that `nopLogger` implements `types.Logger` ? 

If so then making it part of that package making it easier for others to consume in future would help reduce code duplication.

File: client/internal/mockserver.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply client/internal/mockserver.go]: I disagree, we should advocate for using first party package instead of trying to use our own.

**Comment by tigrannajaryan** [comment]
@MovieStoreGuy Please let me know if any of the outstanding comments you have are blockers, if not I will merge this PR and we can address them later.

File: client/clientimpl.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go]: I see. I don't know if anything else needs to use it. Let's keep as is and unexported for now. Can move if needed later.

File: client/internal/mockserver.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/mockserver.go]: I am confused about what you disagree with. Do you disagree that I changed it to use the package that you suggested to use? Or do you disagree that there is a helper MockServer that wraps httptest.NewServer?

File: client/internal/mockserver.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply client/internal/mockserver.go:39]: Sorry, I didn't read this was a test.

The idea I had in mind was to upgrade all connections to a websocket if possible within the middleware instead of doing it per endpoint.

**Comment by MovieStoreGuy** [comment]
No worries, just going through them now :) 

Say give me 30m?

File: client/clientimpl.go [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+	// OpAMP Server URL.
+	url *url.URL
+
+	// HTTP request headers to use when connecting to OpAMP Server.
+	requestHeader http.Header
+
+	// Websocket dialer and connection.
+	dialer    websocket.Dialer
+	conn      *websocket.Conn
+	connMutex sync.RWMutex
```
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply client/clientimpl.go:40]: This might be a practice in my org but I had seen it referred to as a Mutex hat https://dmitri.shuralyov.com/idiomatic-go#mutex-hat

File: client/clientimpl.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply client/clientimpl.go:148]: I am just concerned that we would be leaving work undone which could lead for a bad time.

File: client/clientimpl.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply client/clientimpl.go:231]: The one I had been using inside some of my projects is `github.com/tilinna/clock` since it can load the clock from context.

File: client/clientimpl.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go:148]: The typical user-case for cancelling Stop is usually the final shutdown of the process. In this case it is likely preferable to make sure Stop() finishes within some reasonable time even if there is work remaining than to risk Stop() blocking the process shutdown infinitely long.
We can carefully review the code to see what can potentially go wrong if Stop() is cancelled (you are right that it may be a concern), but I think it is highly desirable to make Stop() cancellable and not block potentially infinitely.
Let's merge the PR and we can see what needs improving here (I feel we need more test coverage to uncover the edge cases anyway).

File: client/clientimpl.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/clientimpl.go:231]: Sounds good.

File: client/internal/mockserver.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/mockserver.go:39]: It probably does not matter much in this test code, but if you see a room for improvement we can certainly do a follow up PR (a PR from you is also welcome).

File: client/internal/callbacks.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (MovieStoreGuy) [reply client/internal/callbacks.go:10]: I think the the things that erks me is the verbosity since this isn't typically I had seen in golang.

I am not sure if this is going to be an object that is passed around or scoped within a stack.

I hadn't checked to see how it is being used but I imagine you could do something like this:

```golang
type Callback func(ctx context.Context) error

type Callbacks []Callback

func OnConnectionResolve(err error) Callback {
    return func(ctx context.Context) error {
         // Do something with error and attempt connection
        return nil
    }
}
```

I can see this struct getting infinitely big with arbitrary function signatures.

**Review by MovieStoreGuy — APPROVED** [review]
The comments I have on this shouldn't stop actively iterating on this.

I think if there is anything, it can be addressed in following PRs

File: client/internal/callbacks.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply client/internal/callbacks.go:10]: I do not see a good, type-safe way to do what you suggest. Perhaps there is a way but I fail to see it. I'll go ahead with the current approach for now to make progress on the implementation but if you can suggest an improvement I will be happy to review it if you submit a PR.

**Comment by tigrannajaryan** [comment]
OK, I am going to merge this. Improvement suggestions and PRs are welcome.
